File: src/org/jetbrains/agda/mixfix/Grammar.java
package org.jetbrains.agda.mixfix;


import com.intellij.psi.PsiElement;
import org.jetbrains.agda.psi.Application;

import java.util.*;

public class Grammar {
    private Map<String, PsiElement> myDeclarations;
    private List<String> myOperatorParts = new ArrayList<String>();

    public Grammar(Map<String, PsiElement> declarations) {
        myDeclarations = declarations;
        for (String declaration: new HashSet<String>(myDeclarations.keySet())) {
            if (!declaration.contains("_")) {
                myDeclarations.remove(declaration);
            }
        }
        Set<String> operationParts = getOperationParts(myDeclarations);
        myOperatorParts.addAll(operationParts);
    }

    public static Set<String> getOperationParts(Map<String, PsiElement> declarations) {
        Set<String> operationParts = new HashSet<String>();

        for (String declaration: declarations.keySet()) {
            if (!declaration.contains("_")) {
                continue;
            }
            for (String str: decode(declaration)) {
                if (!str.equals("_")) {
                    operationParts.add(str);
                }
            }
        }
        return operationParts;
    }

    private static List<PsiElement> getListOfTerminals(Application application) {
        while (application.getParent() instanceof Application) {
           application = (Application)application.getParent();
        }
        List<PsiElement> result = new ArrayList<PsiElement>();
        PsiElement element = application;
        while (element instanceof Application) {
            PsiElement[] expressionList = element.getChildren();
            result.add(expressionList[0]);
            if (expressionList.length > 1) {
                element = expressionList[1];
            }  else {
                break;
            }
        }
        result.add(element);
        return result;
    }

    public static TreeElement parse(Application application) {
        List<PsiElement> listOfTerminals = getListOfTerminals(application);
        Map<String, PsiElement> declarations = org.jetbrains.agda.scope.namespace.getGlobalDeclarations(application);

        return new Grammar(declarations).parse(listOfTerminals);
    }

    private TreeElement parse(List<PsiElement> listOfTerminals) {
        List<TreeElement> list = new ArrayList<TreeElement>();
        for (PsiElement element : listOfTerminals) {
            list.add(new TreeElement(element, null));
        }
        TreeElement term = parseIterative(parseApplications(list));
        return term;
    }

    private List<TreeElement> parseApplications(List<TreeElement> list) {
        List<TreeElement> result = new ArrayList<TreeElement>();
        TreeElement current = null;
        for (TreeElement term : list) {
            if (myOperatorParts.contains(term.getText())) {
                if (current != null) {
                    result.add(current);
                    current = null;
                }
                result.add(term);
            } else {
                if (current == null) {
                    current = new TreeElement(null, null, term);
                } else {
                    current = new TreeElement(null, null, current, new TreeElement(null, null, term));
                }
            }
        }
        if (current != null) {
            result.add(current);
        }
        return result;
    }

    private TreeElement parseIterative(List<TreeElement> list) {
        if (list.size() == 1) {
            return list.get(0);
        }
        boolean good = true;
        for (TreeElement treeElement : list) {
            if (myOperatorParts.contains(treeElement.getText())) {
                good = false;
                break;
            }
        }
        if (good) {
            TreeElement current = null;
            for (TreeElement treeElement : list) {
                if (current == null) {
                    current = treeElement;
                } else {
                    current = new TreeElement(null, null, current, treeElement);
                }
            }
            return current;
        }
        for (int i = 0; i < list.size(); i++) {
            TreeElement term = applyRule(list, i);
            if (term != null) {
                return term;
            }
        }
        return null;
    }

    private TreeElement applyRule(List<TreeElement> list, int i) {
        for (String rule: myDeclarations.keySet()) {
            String[] ruleCode = decode(rule);
            int length = ruleCode.length - 1;
            if (i >= length) {
                boolean matches = true;
                for (int j = 0; j < ruleCode.length; j++) {
                    if (ruleCode[j].equals("_")) {
                        if(myOperatorParts.contains(list.get(i + j - length).getText())) {
                            matches = false;
                            break;
                        }
                    } else {
                        if(!ruleCode[j].equals(list.get(i + j - length).getText())) {
                            matches = false;
                            break;
                        }
                    }
                }

                if (matches) {
                    List<TreeElement> termList = list.subList(i - length, i + 1);
                    TreeElement term = new TreeElement(null, myDeclarations.get(rule), termList.toArray(new TreeElement[termList.size()]));

                    return parseIterative(replace(list, i, length, term));
                }
            }
        }
        return null;
    }

    private List<TreeElement> replace(List<TreeElement> list, int i, int size, TreeElement v) {
        ArrayList<TreeElement> terms = new ArrayList<TreeElement>(list);
        for (int j = 0; j <= size; j++) {
            terms.remove(i - size);
        }
        terms.add(i - size, v);
        return terms;
    }

    private static String[] decode(String rule) {
        if (rule.contains("_")) {
            List<String> result = new ArrayList<String>();
            int start = 0;
            for (int i = 0; i < rule.length(); i++) {
                if (rule.charAt(i) == '_') {
                    if (start < i) {
                        result.add(rule.substring(start, i));
                    }
                    result.add("_");
                    start = i + 1;
                } else {
                    if ((i + 1) == rule.length() ) {
                        result.add(rule.substring(start, i + 1));
                    }
                }
            }
            return result.toArray(new String[result.size()]);
        } else {
            return new String[]{rule};
        }
    }
}

File: src/org/jetbrains/agda/mixfix/Grammar.kt
package org.jetbrains.agda.mixfix

import com.intellij.psi.PsiElement
import java.util.ArrayList
import java.util.HashSet
import org.jetbrains.agda.scope.AgdaExpressionScope
import org.jetbrains.agda.psi.Application
import java.util.Collections
import java.util.Arrays
import org.jetbrains.agda.psi.FqName


public open class Grammar(val myDeclarations : MutableMap<String, PsiElement>) {
    private val myOperatorParts = ArrayList<String>()

    private open fun parse(listOfTerminals : List<PsiElement>) : TreeElement? {
        val list = ArrayList<TreeElement>()
        for (element : PsiElement in listOfTerminals) {
            val isOperation = (element is FqName) && myOperatorParts.contains(element.getText())
            list.add(TermElement(element, isOperation))
        }
        return parseIterative(parseApplications(list))
    }

    private open fun parseApplications(list : List<TreeElement>) : List<TreeElement> {
        val result = ArrayList<TreeElement>()
        var current : TreeElement? = null
        for (term in list) {
            if (myOperatorParts.contains(term.getText())) {
                if (current != null) {
                    result.add(current!!)
                    current = null
                }
                result.add(term)
            } else {
                if (current == null) {
                    current = term
                } else {
                    current = ApplicationTreeElement(current!!, term)
                }
            }
        }
        if (current != null) {
            result.add(current!!)
        }

        return result
    }

    private open fun parseIterative(list : List<TreeElement>) : TreeElement? {
        if ((list.size()) == 1) {
            return list.get(0)
        }

        var good : Boolean = true
        for (treeElement in list) {
            if (myOperatorParts.contains(treeElement.getText())) {
                good = false
                break
            }
        }
        if (good) {
            var current : TreeElement? = null
            for (treeElement in list) {
                if (current == null) {
                    current = treeElement
                } else {
                    current = ApplicationTreeElement(current!!, treeElement)
                }
            }
            return current
        }

        for (i in 0..list.size() - 1) {
            val term = applyRule(list, i)
            if (term != null) {
                return term
            }

        }
        return null
    }
    private open fun applyRule(list : List<TreeElement>, i : Int) : TreeElement? {
        for (rule : String? in myDeclarations.keySet()) {
            if (rule?.equals("_")!!) {
                continue
            }

            val ruleCode : List<String> = decode(rule!!)
            val length : Int = (ruleCode.size) - 1
            if (i >= length) {
                var matches : Boolean = true
                for (j in 0..ruleCode.size - 1) {
                    if (ruleCode[j].equals("_")) {
                        if (myOperatorParts.contains(list.get(i + j - length).getText())) {
                            matches = false
                            break
                        }
                    } else {
                        if (!ruleCode[j].equals(list.get(i + j - length).getText())) {
                            matches = false
                            break
                        }
                    }
                }
                if (matches) {
                    val termList = list.subList(i - length, i + 1)
                    val term  = OperationElement(myDeclarations.get(rule)!!, termList)
                    return parseIterative(replace(list, i, length, term))
                }

            }

        }
        return null
    }

    private open fun replace(list : List<TreeElement>, i : Int, size : Int, v : TreeElement) : List<TreeElement> {
        var terms = ArrayList<TreeElement>(list)
        for (j in 0..size) {
            terms.remove(i - size)
        }
        terms.add(i - size, v)
        return terms
    }

    {
        for (declaration : String? in HashSet<String>(myDeclarations.keySet())) {
            if (!declaration!!.contains("_")) {
                myDeclarations.remove(declaration)
            }

        }
        val operationParts : Set<String> = getOperationParts(myDeclarations)
        myOperatorParts.addAll(operationParts)
    }

    class object {
        public open fun getOperationParts(declarations : Map<String, PsiElement>) : Set<String> {
            var operationParts : MutableSet<String> = HashSet<String>()
            for (declaration in declarations.keySet()) {
                if (!declaration.contains("_")) {
                    continue
                }

                for (str : String? in decode(declaration)) {
                    if (!str!!.equals("_")) {
                        operationParts.add(str)
                    }

                }
            }
            return operationParts
        }
        private open fun getListOfTerminals(var application : Application) : List<PsiElement>? {
            while ((application.getParent() is Application)) {
                application = (application.getParent() as Application)
            }
            val result = ArrayList<PsiElement>()
            var element : PsiElement = application
            while (element is Application) {
                val expressionList : Array<PsiElement> = (element : PsiElement).getChildren()
                result.add(expressionList[0])
                if ((expressionList.size) > 1) {
                    element = expressionList[1]
                } else {
                    break
                }
            }
            result.add(element)
            return result
        }

        public open fun parse(application : Application) : TreeElement? {
            val listOfTerminals : List<PsiElement>? = getListOfTerminals(application)
            val scope = AgdaExpressionScope(application)
            val declarations : MutableMap<String, PsiElement> = scope.getVisibleDeclarations()
            return Grammar(declarations).parse(listOfTerminals!!)
        }

        private open fun decode(rule : String) : List<String> {
            if (rule.contains("_")) {
                var result = ArrayList<String>()
                var start : Int = 0
                for (i in 0..rule.length() - 1) {
                    if ((rule.charAt(i)) == '_') {
                        if (start < i) {
                            result.add(rule.substring(start, i))
                        }

                        result.add("_")
                        start = i + 1
                    }
                    else
                    {
                        if ((i + 1) == rule.length())
                        {
                            result.add(rule.substring(start, i + 1))
                        }

                    }
                }
                return result
            } else {
                return Collections.singletonList(rule)
            }
        }
    }
}
